src/canvas/Canvas-events.spec.ts

```diff
@@ -0,0 +1,1908 @@
+    canvas.__onMouseUp(e);
+  });
+
+  it('handles activeOn object selection', () => {
+    const rect = new Rect({ width: 200, height: 200, activeOn: 'down' });
+    canvas.add(rect);
+    const e = {
+      clientX: 30,
+      clientY: 15,
+      which: 1,
+      target: canvas.upperCanvasEl,
+    };
+
+    // @ts-expect-error private method
+    canvas.__onMouseDown(e);
+    expect(
+      canvas._activeObject,
+      'with activeOn of down object is selected on mouse down',
+    ).toBe(rect);
+
+    // @ts-expect-error private method
+    canvas.__onMouseUp(e);
+    canvas.discardActiveObject();
+    rect.activeOn = 'up';
+
+    // @ts-expect-error private method
+    canvas.__onMouseDown(e);
+    expect(
+      canvas._activeObject,
+      'with activeOn of up object is not selected on mouse down',
+    ).toBeUndefined();
+
+    // @ts-expect-error private method
+    canvas.__onMouseUp(e);
+    expect(
+      canvas._activeObject,
+      'with activeOn of up object is selected on mouse up',
+    ).toBe(rect);
+  });
+
+  it('handles specific bug #5317 for multiselection', () => {
+    const greenRect = new Rect({
+      width: 300,
+      height: 300,
+      left: 0,
+      top: 0,
+      fill: 'green',
+      selectable: false,
+    });
+    canvas.add(greenRect);
+
+    // add red, half-transparent circle
+    const redCircle = new Circle({
+      radius: 40,
+      left: 200,
+      top: 100,
+      fill: 'red',
+      opacity: 0.5,
+    });
+    canvas.add(redCircle);
+
+    // add blue, half-transparent circle
+    const blueCircle = new Circle({
+      radius: 40,
+      left: 0,
+      top: 0,
+      fill: 'blue',
+      opacity: 0.5,
+    });
+    canvas.add(blueCircle);
+
+    const e = {
+      clientX: 40,
+      clientY: 40,
+      which: 1,
+      target: canvas.upperCanvasEl,
+    };
+    // @ts-expect-error private method
+    canvas.__onMouseDown(e);
+    expect(
+      canvas._activeObject,
+      'blue circle is selected with first click',
+    ).toBe(blueCircle);
+
+    // @ts-expect-error private method
+    canvas.__onMouseUp(e);
+    const e2 = {
+      clientX: 240,
+      clientY: 140,
+      which: 1,
+      target: canvas.upperCanvasEl,
+      shiftKey: true,
+    };
+    // @ts-expect-error private method
+    canvas.__onMouseDown(e2);
+
+    const selection = canvas.getActiveObjects();
+    expect(selection[1], 'blue circle is still selected').toBe(blueCircle);
+    expect(selection[0], 'red circle is selected with shift click').toBe(
+      redCircle,
+    );
+
+    // @ts-expect-error private method
+    canvas.__onMouseUp(e2);
+    const e3 = {
+      clientX: 140,
+      clientY: 90,
+      which: 1,
+      target: canvas.upperCanvasEl,
+      shiftKey: true,
+    };
+
+    canvas.on('mouse:down', function (options) {
+      // TODO: fix this, for some reason first target on mouse down is active selection and then second target is the green rectangle
+      if (options.target instanceof ActiveSelection) {
```

Comment:
@asturur I don't know how this test worked in qunit but in both jsdom and in the browser the first event that comes here is Active selection and then the green box 
I guess you will have to have a look here since you have far more knowledge than I do when it comes to fabric source code

for now I workarounded this with a return if it is now a green box but clearly this is not a solution